### PR TITLE
Update GEP-1762 with Gateway name limit recommendations

### LINKED ISSUE #2592: GEP-1762 needs clarification on long Gateway names > 63 characters
**What happened**:

Prior to GEP-1762, Contour's Gateway provisioner has been adding a label to resources to denote which Gateway "owns" it: `projectcontour.io/owning-gateway-name` set to the Gateway name.

While reviewing https://github.com/kubernetes-sigs/gateway-api/pull/2582 with Contour and also implementing the Gateway infrastructure labels feature, we realized that Gateway names longer than 63 characters cause failures to create resources generated with this label.

While this can be solved for Contour's implementation-specific labels with a hash or some other solution, this is not exclusive to Contour, the upstream GEP-1762 also has requirements that may cause many implementations to run into this. Also, including a conformance test that exhibits this may be problematic for some implementations.

@skriss pointed out as well that GEP-1762 makes this same type of label, including the "owning" Gateway as the value, required for implementations that generate deployment resources dynamically: https://gateway-api.sigs.k8s.io/geps/gep-1762/#automated-deployments

**What you expected to happen**:

GEP-1762 should be amended to describe what to do with the `gateway.networking.k8s.io/gateway-name` label when dealing with a Gateway with a long name.

**How to reproduce it (as minimally and precisely as possible)**:

Run the conformance test `HTTPRouteInvalidParentRefSectionNameNotMatchingPort` (on `main`) with Contour, or create a Gateway with a name longer than 63 characters and have Contour reconcile it.

**Anything else we need to know?**:

Hashing a long Gateway name and setting that as the value of may be an answer, however that means we will lose the immediate human readability of this feature.

Another solution would be to limit Gateway names to < 63 characters and longer names would not be Accepted by implementations.

### PR DESCRIPTION
**Pull request by sunjayBhatia**
**What type of PR is this?**

/kind cleanup
/kind documentation
/kind gep

**What this PR does / why we need it**:

Updates GEP-1762 to provide guidance for implementers on what to do with the `gateway.networking.k8s.io/gateway-name` (to be put on generated in-cluster resources) when a Gateway's name exceeds 63 characters (the label value length limit).

Also updates Gateway resource to recommend users create Gateways with <= 63 character names.

**Which issue(s) this PR fixes**:

Fixes #2592 

**Does this PR introduce a user-facing change?**:

```release-note
Gateway names are recommended to be no longer than the maximum label value length of 63 characters to ensure compatibility with in-cluster implementations which set the Gateway name as a label on generated resources.
```

### CODE REVIEW

**Review by arkodg — APPROVED** [review]
/approve

thanks !

File: apis/v1/gateway_types.go
Comment (robscott) [inline apis/v1/gateway_types.go:37]: It seems like 63 is too long if 63 is the max length of some of the values we're trying to populate. Ideally wouldn't you want to translate `foo-gateway` to a generated `contour-foo-gateway` or similar? I'm also not sure if this belongs in the API spec or should just be in more user-facing guidelines documentation for using the API.

File: geps/gep-1762/index.md
Comment (robscott) [inline geps/gep-1762/index.md:62]: Why not just use Kubernetes `generateName` field? If we need to predictably tie this back to a Gateway, we already have the label defined above. I'm guessing we want to avoid stomping over any preexisting resources with generated ones, and anything too prescriptive here would effectively _require_ that behavior. On the other hand, `generateName` ensures that we always get a unique + new name and we can rely on label selection for reliable lookups instead of name.

**Review by robscott — COMMENTED** [review]
Thanks @sunjayBhatia!

File: geps/gep-1762/index.md
Comment (sunjayBhatia) [reply geps/gep-1762/index.md:62]: hm, not sure how the `generateName` field helps here? This update isn't for specifying the name of generated resources but what to do with the `gateway.networking.k8s.io/gateway-name` label value (which should be set to the Gateway name) in the case a Gateway's name is > 63 characters

File: apis/v1/gateway_types.go
Comment (sunjayBhatia) [reply apis/v1/gateway_types.go:37]: 63 is the max label value length, this shouldn't have any immediate bearing on generated resource names

Comment (sunjayBhatia) [reply apis/v1/gateway_types.go:37]: I can move this to another guide instead though if that is preferred, maybe better here: https://gateway-api.sigs.k8s.io/api-types/gateway/

File: geps/gep-1762/index.md
Comment (robscott) [reply geps/gep-1762/index.md:62]: 🤦 I definitely did not read this clearly enough, sorry for the confusion!

File: apis/v1/gateway_types.go
Comment (robscott) [reply apis/v1/gateway_types.go:37]: >63 is the max label value length, this shouldn't have any immediate bearing on generated resource names

🤦  did not read this closely enough, nevermind.

>I can move this to another guide instead

We don't really have any strong guidance here as far as where user-facing tips like this belongs. Generally we reserve RFC language like SHOULD and MUST in the API Spec for _implementation_ guidance, but it also feels weird to have "It is recommended" instead of "SHOULD" here. I can't really think of better solutions here, but interested in what @youngnick thinks.

**Review by robscott — COMMENTED** [review]
Thanks @sunjayBhatia! Let's give a few days for @youngnick to look at this in case he has better ideas for where this guidance could go. If not, I'm fine with some lazy consensus where we can just remove the hold sometime next week if no strong opinions emerge.

/lgtm
/approve
/hold for second opinion

**Comment by sunjayBhatia** [comment]
> Thanks @sunjayBhatia! Let's give a few days for @youngnick to look at this in case he has better ideas for where this guidance could go. If not, I'm fine with some lazy consensus where we can just remove the hold sometime next week if no strong opinions emerge.
> 
> /lgtm /approve /hold for second opinion

any further thoughts on this @robscott @youngnick

File: apis/v1/gateway_types.go
Comment (youngnick) [reply apis/v1/gateway_types.go:37]: Yeah, this is a good use case for SHOULD - in that we would prefer implementations be aware of the complexities here as well.

Let's change "It is recommended..." to "The Gateway name field SHOULD be no longer than 63 characters."

**Comment by youngnick** [comment]
Yeah, I like these changes, with the small exception of moving to SHOULD - RFC language is clearer for both implementations and users.

**Comment by youngnick** [comment]
LGTM with the SHOULD changes.

**Comment by mlavacca** [comment]
@sunjayBhatia the PR is in pretty advanced state, are you still interested in moving this forward?

**Pull request closed without merging**
